src/poetry/puzzle/provider.py

```diff
@@ -454,15 +458,33 @@ def _search_for_url(self, dependency: URLDependency) -> Package:
         return package
 
     @classmethod
-    def get_package_from_url(cls, url: str) -> Package:
+    def get_package_from_url(
```

Comment:
The function still allows `None` because of `poetry/console/commands/init.py`. I don't exactly know when that `PyProjectException` can be raised (I just copied it from a few lines above). If the exception can never be raised (but I doubt) then I would gladly simplify the codebase and remove the tempdir usage

Reply:
`PyProjectException` is raised if there is an issue with `pyproject.toml`. I think you can remove the exception handling for `ArtifactCache` in `init.py`.

Reply:
Several tests of `init` where indeed failing when moving the artifact_cache fetching out of the try/except, so I went for keeping it

Reply:
Ah, that's because there may be no pyproject.toml when running `poetry init`. I handled that as well.